lib/internal/streams/async_iterator.js

```diff
@@ -118,31 +153,30 @@ const ReadableStreamAsyncIteratorPrototype = ObjectSetPrototypeOf({
   },
 
   return() {
-    return new Promise((resolve, reject) => {
-      const stream = this[kStream];
-
-      // TODO(ronag): Remove this check once finished() handles
-      // already ended and/or destroyed streams.
-      const ended = stream.destroyed || stream.readableEnded ||
-        (stream._readableState && stream._readableState.endEmitted);
-      if (ended) {
-        resolve(createIterResult(undefined, true));
-        return;
-      }
+    return finish(this);
+  },
 
-      finished(stream, (err) => {
-        if (err && err.code !== 'ERR_STREAM_PREMATURE_CLOSE') {
-          reject(err);
-        } else {
-          resolve(createIterResult(undefined, true));
-        }
-      });
-      stream.destroy();
-    });
+  throw(err) {
+    return finish(this, err);
   },
 }, AsyncIteratorPrototype);
 
 const createReadableStreamAsyncIterator = (stream) => {
+  if (typeof stream.read !== 'function') {
+    // v1 stream
+
+    if (!PassThrough) {
+      PassThrough = require('_stream_passthrough');
+    }
+
+    const src = stream;
+    stream = new PassThrough({ objectMode: true });
+    src
+      .on('error', (err) => stream.destroy(err))
+      .pipe(stream);
+    finished(stream, (err) => destroy(src, err));
+  }
```

Comment:
I'm not sure this is needed. We have `Readable.wrap()` to wrap a v1 stream into a Readable. That _should_ work.

Reply:
not quite... The only requirement here is that the stream has a `pipe` method. Using `wrap` only works with old streams that have `_readableState`.  I don't really see any advantage of using it over the above. I at least would feel more comfortable with it as is and I'm worried about edge cases with `wrap`. Haven't really wrapped my head over `wrap`.

Reply:
I don't see much of an advantage... the wrap version would look like:

```js
    const src = stream;
    stream = new Readable({ objectMode: true });
    stream
      .wrap(src);
      .on('error', (err) => stream.destroy(err));
    finished(stream, (err) => destroy(src, err));
```

`wrap` doesn't properly handle error propagation and `destroy` and still needs some extra help.

Reply:
Actually, I'm not sure `wrap` does the correct thing some edge cases. I could dive into `wrap` if you would prefer that, but I personally think it's better like this.